lib/mongo/error/invalid_handshake_document.rb

```diff
@@ -0,0 +1,38 @@
+# Copyright (C) 2015-2016 MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Mongo
+  class Error
+
+    # This exception is raised when the metadata document sent to the server
+    #   at the time of a connection handshake is invalid.
+    #
+    # @since 2.4.0
+    class InvalidHandshakeDocument < Error
+
+      # Instantiate the new exception.
+      #
+      # @example Create the exception.
+      #   InvalidHandshakeDocument.new(app_name)
+      #
+      # @param [ String ] app_name The application name option.
+      #
+      # @since 2.4.0
+      def initialize(app_name)
+        super("The provided application name '#{app_name}' cannot exceed 128 bytes " +
+                "and the entire metadata document cannot exceed 512 bytes.")
```

Comment:
The app_name issue is something the user can do something about, but the metadata document issue is not.  As mentioned above, the driver has to truncate so that the document fits in 512 bytes.